### PR TITLE
ENH: Open Prinables.com Links and Zip Archives

### PR DESCRIPTION
**Pull request by Ocraftyone**
Allows Orca to use the "Open in PrusaSlicer" buttons on Printables.com.

Also implements zip archive support since this is needed to download entire projects from Prinatbles.

### CODE REVIEW

**Comment by Ocraftyone** [comment]
Currently this does NOT register with the OS to indicate this ability. It just allows Orca to parse and download when given the same link that would normally be given to PS. Should Orca be registered or should this? I am going back and forth on whether to...

**Comment by Eldenroot** [comment]
3mf is automatically opened by Bambu lab. Not sure for .zip, it should not be touched.

Also there should be some popup confirmation when you open orca for the first time - do you wanna associate... yes/no

**Comment by Ocraftyone** [comment]
Sorry. I didn't make it clear. I mean for the web links, not zip files

**Comment by SoftFever** [comment]
> Currently this does NOT register with the OS to indicate this ability. It just allows Orca to parse and download when given the same link that would normally be given to PS. Should Orca be registered or should this? I am going back and forth on whether to...

You'da man!
I was considering this feature for both Printable and Makerworld recently too.

I thought we have to register Orca to that MIME in the OS, or am I wrong?
Otherwise how does the browser/OS know it need to invoke Orca?

More thoughts regarding the behavior: I'm thinking we should add a menu item so that user can register OrcaSlicer to that MIME type. What do you think?

**Comment by Ocraftyone** [comment]
At the moment you can also manually specify which app to open a link with, which is how I have been testing it

**Comment by Eldenroot** [comment]
Windows - yes, but some SW ask you if you wanna use them and associate them with specific extensions... anyway this implementation could be done in future.

**Comment by Ocraftyone** [comment]
I am looking into it. I have the commit that PS implemented the feature on for reference. At the moment, the trouble is that Orca inherits a downloader feature from BBL which is very different from the PS implementation. So I don't know if I should use PS implementation or keep with the BBL implementation and hack together a solution. @SoftFever do you have a preference?

**Comment by Eldenroot** [comment]
It depends - maybe some hybrid solution to have an easier life when BS will bring new things on their website or new builds will be synced.

I would like to see this in the next nightly ;)

**Comment by Ocraftyone** [comment]
If I went with the PS implementation, I would implement it in a way that it would only effect PS links. The rest would use the existing method.

**Comment by SoftFever** [comment]
> If I went with the PS implementation, I would implement it in a way that it would only effect PS links. The rest would use the existing method.

I would opt for this as well.
Would be great to have different functions/classes handle different links.
Which will make future maintenance work easier.

**Comment by Eldenroot** [comment]
@Ocraftyone any progress with this? To make our life easier .)

**Comment by Ocraftyone** [comment]
Yup. Just took a bit of a break but I am going to implement the rest tonight.

**Comment by Ocraftyone** [comment]
Update to those following: I have been working on this in my off time. I have the PS downloader mostly implemented how I like it. I am working on putting the final touches on the new download url notification and the icons associated with it. I believe that the macOS url registration has been set up and I have located the part that controls the Linux desktop integration. All that remains for url registration is to find the code that adds the link registration to the windows registry and implement link registration on Linux and windows. I would also like to implement a setting that deleted downloaded files after being imported.

**Comment by Eldenroot** [comment]
Finished? I will try if yes ;)

**Comment by Ocraftyone** [comment]
Finished with the windows and macOS implementation. Linux is WIP

**Comment by Ocraftyone** [comment]
@SoftFever The implementation is basically done. Only question I have is whether I should fix Linux desktop integration within this patch. Without fixing integration, the feature will be basically broken on Linux. I am planning on fixing this and hopefully modernizing how file association is handled on windows in a PR after this is merged.

**Comment by SoftFever** [comment]
> @SoftFever The implementation is basically done. Only question I have is whether I should fix Linux desktop integration within this patch. Without fixing integration, the feature will be basically broken on Linux. I am planning on fixing this and hopefully modernizing how file association is handled on windows in a PR after this is merged.
> 
> Other than the above mentioned issue, this implementation is finished and ready for testing. Windows seems to be good. macOS could use a once over to make sure it is good.

Thank you so much!
I will start to review and test it. It may take some time

**Comment by SoftFever** [comment]
@Ocraftyone 
I tried it on Mac, it seems didn't work.
I have this option ticked: 
Screenshot 2024-03-31 at 12 28 24 PM
When click the button on Printables.com, it still invokes PS.
Looks like the registration step is missing?

**Comment by Ocraftyone** [comment]
I believe the registration step is there, but it is done statically via the Plist.info file. I don't have access to a mac to do any investigation, but I wonder if there is a way to set which application handles the url

**Comment by SoftFever** [comment]
> I believe the registration step is there, but it is done statically via the Plist.info file. I don't have access to a mac to do any investigation, but I wonder if there is a way to set which application handles the url
> 
> Edit: Nope! Apparently it isn't a GUI, but a file located at ~/Library/Preferences/com.apple.LaunchServices/com.apple.launchservices.secure.plist in an array called LSHandlers

I see.
I will play around with the current implement then share more thoughts.
Statically replace other application's association might not be the proper way. We won't be able to do this eventually.
I will share more when I have better understanding of the open file thing mechanism.

**Comment by nevack** [comment]
> I believe the registration step is there, but it is done statically via the Plist.info file. I don't have access to a mac to do any investigation, but I wonder if there is a way to set which application handles the url

This can be done programmatically at runtime.
If you don't mind, I can submit a patch here.

I have already done something similar for Transmission [see](https://github.com/transmission/transmission/pull/6120/files#diff-e07c527e3ccacbe34fd0f3864d6073c2c29d28c29a5f7909f8a43192100f6eb6)

**Comment by nevack** [comment]
@SoftFever Could you guide me where is the proper place to write mac platform specific code?
Is there already common place for ObjC/ObjC++?
Or should we use plain C macOS APIs?

**Comment by nevack** [comment]
@Ocraftyone Basically, you need to execute something like this when "checkbox" is checked
https://github.com/transmission/transmission/pull/6120/files#diff-e07c527e3ccacbe34fd0f3864d6073c2c29d28c29a5f7909f8a43192100f6eb6R150-R180

After this you can't just "give up" being the default app for url scheme, so the checkbox should be deactivated.
It will be reactivated only when any other app asks to be default.

**Comment by nevack** [comment]
> When click the button on Printables.com, it still invokes PS.
> Looks like the registration step is missing?

Yep, currently OrcaSlicer only declares it can handle such url scheme, but does not set itself as default.
You can play with https://github.com/relikd/URL-Scheme-Defaults until this is implemented in code.

File: src/platform/osx/Info.plist.in [hunk 1]
```diff
@@ -121,6 +121,17 @@
       <string>Alternate</string>
     </dict>
   </array>
+  <key>CFBundleURLTypes</key>
```
Comment (nevack) [inline src/platform/osx/Info.plist.in:124]: There's already `CFBundleURLTypes` block above.
You should merge this into that one.

Comment (nevack) [reply src/platform/osx/Info.plist.in:124]: Also there's `cmake/modules/MacOSXBundleInfo.plist.in`, which should be also updated.

**Comment by nevack** [comment]
I think this feature can be rolled out gradually - at first only for Windows.
Then we can add support for macOS and Linux.
I think I can do proper impl for macOS.

Also, can you provide any `prusaslicer://` url to test this with?

**Comment by Ocraftyone** [comment]
> @Ocraftyone Basically, you need to execute something like this when "checkbox" is checked
> https://github.com/transmission/transmission/pull/6120/files#diff-e07c527e3ccacbe34fd0f3864d6073c2c29d28c29a5f7909f8a43192100f6eb6R150-R180
> 
> After this you can't just "give up" being the default app for url scheme, so the checkbox should be deactivated.
> It will be reactivated only when any other app asks to be default. 

We should leave the check box active. Disabling it prevents links from being opened, even if the url is registered.

**Comment by Ocraftyone** [comment]
> I think this feature can be rolled out gradually - at first only for Windows.
> Then we can add support for macOS and Linux.
> I think I can do proper impl for macOS.
> 
> Also, can you provide any `prusaslicer://` url to test this with?

This sounds good. I have some tweaking I want to do to both windows and Linux on how they register file types and urls

**Comment by Ocraftyone** [comment]
> @SoftFever Could you guide me where is the proper place to write mac platform specific code?
> Is there already common place for ObjC/ObjC++?
> Or should we use plain C macOS APIs?

Platform specific code is typically done via c++ definitions inline and I don't believe objective c(++) is used in this project. Currently, windows is the only platform with working dynamic file type association. This is done with two functions in GUI_App. associate_url and disassociate_url.

**Comment by Eldenroot** [comment]
How is it handled via Prusa Slicer?

**Comment by Ocraftyone** [comment]
PrusaSlicer doesn't handle it because they assume only they will handle the PrusaSlicer link type

**Comment by nevack** [comment]
> It would be better to ask the user to explicitly register/deregister it (through a menu item).

I also think it's a bad practice to register itself as default handler unconditionally without user interaction.
Maybe a dedicated checkbox in main settings window will do?

Basically, there should be API like this to make it work.
```cpp
bool is_set_default();
void set_as_default();
void unset_as_default();
```
Maybe, in a separate "interface" (pure abstract C++ class).

This APIs should be implemented for each supported platform

Also as a side note, I would like this to be merged sooner than later, to start iterating on macOS impl part.

**Comment by nevack** [comment]
@Ocraftyone @SoftFever If you are okay with this, I propose to remove any macOS modifications for now (`Info.plist` at least). These changes are
1. Insufficient.
2. Not made properly.

Let's not advertise this url scheme handler to OS until it's properly implemented

**Comment by Ocraftyone** [comment]
> @Ocraftyone It works on Windows 👍 
> 
> I'm thinking the registration behavior can be tweaked a bit.
> Currently, it automatically registers the current instance on startup. It would be better to ask the user to explicitly register/deregister it (through a menu item).
> 
> 

Yeah. This is PS default behavior and how file types are registered in both. I was wanting to look into using the windows settings to handle file type registration.

**Comment by nevack** [comment]
@Ocraftyone JFYI
https://github.com/prusa3d/PrusaSlicer/releases/tag/version_2.7.4
https://github.com/prusa3d/PrusaSlicer/compare/version_2.7.3...version_2.7.4

**Comment by Ocraftyone** [comment]
> @Ocraftyone JFYI
> https://github.com/prusa3d/PrusaSlicer/releases/tag/version_2.7.4
> https://github.com/prusa3d/PrusaSlicer/compare/version_2.7.3...version_2.7.4
> 

Thanks. I'll take a deeper look when I get a chance

**Comment by SoftFever** [comment]
> > It would be better to ask the user to explicitly register/deregister it (through a menu item).
> 
> I also think it's a bad practice to register itself as default handler unconditionally without user interaction. Maybe a dedicated checkbox in main settings window will do?
> 
> Basically, there should be API like this to make it work.
> 
> ```c++
> bool is_set_default();
> void set_as_default();
> void unset_as_default();
> ```
> 

Would be better to register/unregister itself per user request (every time) instead of any kind of implicit register action, not even . Checkbox in main setting is what @Ocraftyone did currently.

There are two issues with such behavior:

1. The setting file is shared by all Orca instances; whenever the user opens Orca, it reads the setting and registers itself. When we have multiple versions of Orca (stable/beta/dev), this causes a mess.
2. When a user doesn't want Orca to handle such file type/URL, a setting is not obvious; one will easily forget that, then there are endless fights for the file type handling between Orca and other software. That's kind of rogue behavior that should be avoided.

**Comment by Ocraftyone** [comment]
@SoftFever This is something I need to address in windows too because it has the same unintuitive behavior. I think if the application thinks it should be registered, it should check the registry if it is indeed the registered application (by comparing the registered application path). if it should be registered but is not, then ask the user for intervention. This wouldn't interfere with updates, because the app would still be in the same place, but if the app is a dev version or a portable version, it wont automatically register.

**Comment by nevack** [comment]
@Ocraftyone I kindly ask you to create some common interface for this functionality, so I can start implementing macOS part independently.

Something like described here https://github.com/SoftFever/OrcaSlicer/pull/3823#issuecomment-2034563240

It seems like we are blocked here, but I don't clearly see why 😄

**Comment by SoftFever** [comment]
> @SoftFever This is something I need to address in windows too because it has the same unintuitive behavior. I think if the application thinks it should be registered, it should check the registry if it is indeed the registered application (by comparing the registered application path). if it should be registered but is not, then ask the user for intervention. This wouldn't interfere with updates, because the app would still be in the same place, but if the app is a dev version or a portable version, it wont automatically register.

Hmm, I still think the right behaviour is 
User click the register button -> check the current registration info and popup a warning dialog about overwrite -> register current instance

**Comment by Ocraftyone** [comment]
@SoftFever
My initial plan to fix up how windows registers file/link types was to use Window's settings dialog.

Currently, Orca and PS forcibly set themselves as the handlers for the various file types on Windows. Instead, we can register the application with windows and specify the different file types it can handle. Windows will then show the application in the file associations section of the settings panel. We could then provide a link to the associations page for Orca in preferences. This would leave orca itself out of the registration process, preventing some of the messy issues it currently creates, removes the outdated registration code, and we no longer need to check if we are indeed the default application. All of that is handled by Windows.

There are a few downsides to this idea. 1) Below Windows 10 won't work (but they aren't supported any more by Microsoft) and 2) portable instances of Orca won't be able to be registered (the registry steps would be handled by the CPack installer)

During creating this PR, I had considered making that change a part of it, but I believe it would be a bit out of scope. My suggestion would be to leave the windows portion alone in this PR since it follows the current association standard, then in a later PR, update to a more modern implementation.

Here is the inspiration for the change. Nanazip (a 7zip fork focused on a modern Windows GUI implementation) uses the idea stated above.

image
image

**Comment by Ocraftyone** [comment]
> @Ocraftyone I kindly ask you to create some common interface for this functionality, so I can start implementing macOS part independently.
> 
> Something like described here [#3823 (comment)](https://github.com/SoftFever/OrcaSlicer/pull/3823#issuecomment-2034563240)
> 
> It seems like we are blocked here, but I don't clearly see why 😄

I believe the reason we are blocked is because until now only Windows has had dynamic registration functionality. On linux and macos, it was registered at install and not touched again. That, and how windows does its registration is very hacky. It just sets it every time an instance of orca starts. We need to determine if/how we should update windows to match the new macOS functionality. In the mean time, I am happy to point you to where registration is currently done in windows if you want to do a dirty implementation that we can refactor later. `GUI_App::open_preferences` opens the preferences dialog and then registration is run after it is closed. [This](https://github.com/Ocraftyone/OrcaSlicer-Ocraftyone/blob/40d31d6245e7ce37483675f832696296d5b6d32b/src/slic3r/GUI/GUI_App.cpp#L2272-L2287) block of code is run on init and registers the application if the config says it should be associated.

**Comment by SoftFever** [comment]
I'm fine with focusing only on Windows for this PR. I will create another PR to modify the behavior. It's very important that if we are modifying system settings, especially those of other software types, it must be initiated by the user's explicit and direct action.

**Comment by Ocraftyone** [comment]
> I'm fine with focusing only on Windows for this PR. I will create another PR to modify the behavior. 

I don't quite understand this statement. Are you wanting to include the above mentioned changes in this PR or are you wanting to create a different PR that addresses the windows issues and merge that first?

> It's very important that if we are modifying system settings, especially those of other software types, it must be initiated by the user's explicit and direct action.

Understood. I definitely think the current behavior is not ideal at all. I am surprised it hasn't been handled by the PS team yet.

**Comment by SoftFever** [comment]
> I don't quite understand this statement. Are you wanting to include the above mentioned changes in this PR or are you wanting to create a different PR that addresses the windows issues and merge that first?

Oh, I meant I will create another PR based on your PR. Once that's ready, I will merge your PR first, then that PR.

> 
> > It's very important that if we are modifying system settings, especially those of other software types, it must be initiated by the user's explicit and direct action.
> 
> Understood. I definitely think the current behavior is not ideal at all. I am surprised it hasn't been handled by the PS team yet with it being this hacky.

👍

**Comment by Ocraftyone** [comment]
> Oh, I meant I will create another PR based on your PR. Once that's ready, I will merge your PR first, then that PR.

Are you planning on going with a register button or one that goes to windows settings?

**Comment by SoftFever** [comment]
> > Oh, I meant I will create another PR based on your PR. Once that's ready, I will merge your PR first, then that PR.
> 
> Are you planning on going with a register button or one that goes to windows settings?

I'm leaning towards a button

**Comment by Ocraftyone** [comment]
> I'm leaning towards a button
I would still suggest updating to the more modern approach that allows for configuration via Windows Settings and maybe adding a register option to the installer

**Comment by SoftFever** [comment]
> @SoftFever any progress on the other PR?

@Ocraftyone do you mind I push changes into this PR directly?
I find the changes are less than I expected.

**Comment by Ocraftyone** [comment]
Go for it!

**Comment by SoftFever** [comment]
> Go for it!

Seems I can't push changes into your branches anymore.
So I created a another PR: https://github.com/SoftFever/OrcaSlicer/pull/5416

**Comment by Ocraftyone** [comment]
Do you want to submit as a PR to my branch and I can merge it?

**Comment by SoftFever** [comment]
> Do you want to submit as a PR to my branch and I can merge it?

Thank you!
Since the new PR ha been created already, let's submit it as two PRs.

**Review by SoftFever — APPROVED** [review]

**Comment by SoftFever** [comment]
@Ocraftyone let me know if you find any issue with the changes there

**Pull request merged**